File: Java-Kotlin/src/main/kotlin/dfs/Max Area of Island.kt
package dfs

/**
 * https://leetcode.com/problems/max-area-of-island/
 * */

fun maxAreaOfIsland(grid: Array<IntArray>): Int {
    var ans = 0
    for (i in grid.indices)
        for (j in grid[0].indices)
            if (grid[i][j] == 1)
                ans = Math.max(dfs(grid, i, j), ans)
    return ans
}

private fun dfs(grid: Array<IntArray>, x: Int, y: Int): Int {
    if (x < 0 || y < 0 || x == grid.size || y == grid[0].size || grid[x][y] == 0)
        return 0
    grid[x][y] = 0
    val left = dfs(grid, x, y - 1)
    val right = dfs(grid, x, y + 1)
    val top = dfs(grid, x - 1, y)
    val bottom = dfs(grid, x + 1, y)
    return left + right + top + bottom + 1
}

File: C++/Depth-first Search/Max Area of Island.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * https://leetcode.com/problems/max-area-of-island/
 */

class Solution {
public:
    unsigned int n, m;

    int maxAreaOfIsland(vector<vector<int>> &grid) {
        n = grid.size(), m = grid[0].size();
        int ans = 0;
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                ans = max(ans, dfs(grid, i, j));
        return ans;
    }

    int dfs(vector<vector<int>> &grid, int x, int y) {
        if (x < 0 || y < 0 || x >= n || y >= m || grid[x][y] == 0) return 0;
        grid[x][y] = 0;
        int left = dfs(grid, x, y - 1);
        int right = dfs(grid, x, y + 1);
        int above = dfs(grid, x - 1, y);
        int under = dfs(grid, x + 1, y);
        return left + right + above + under + 1;
    }
};